### PR TITLE
Wrap binary file

### PR DESCRIPTION
**Pull request by mhvk**
Just to show how it looks wrapped - I think I like it quite a bit better, as it gives a cleaner "separation of concerns". Obviously, some of the wrapping could be done in the opener (where I used to have it, sigh...).

### CODE REVIEW

File: baseband/gsb/base.py [hunk 1]
```diff
@@ -56,31 +56,38 @@ class GSBFileReader(VLBIFileBase):
...
     Adds ``read_payload`` method to the basic VLBI binary file wrapper.
-    """
 
-    def read_payload(self, payloadsize, nchan=1, bps=4, complex_data=False):
+    Parameters
+    ----------
+    payloadsize : int
+        Number of bytes to read.
+    nchan : int
+        Number of channels in the data.  Default: 1.
```
Comment (cczhu) [inline baseband/gsb/base.py]: In line with new docstrings:
```
nchan : int, optional
    Number of channels.  Default: 1
```

File: baseband/gsb/base.py [hunk 2]
```diff
@@ -56,31 +56,38 @@ class GSBFileReader(VLBIFileBase):
...
+    bps : int
+        Number of bits per sample (or real/imaginary component).
+        Default: 4.
```
Comment (cczhu) [inline baseband/gsb/base.py]:
```
bps : int, optional
    Bits per elementary sample; i.e. per real or imaginary component for
    complex data.  Default: 4.
```
Not that I'm in love with that description, but having everything the same will make it a cinch to replace if we feel like it.

File: baseband/gsb/base.py [hunk 3]
```diff
@@ -56,31 +56,38 @@ class GSBFileReader(VLBIFileBase):
...
+    complex_data : bool
+        Whether data is complex or float.  Default: False.
```
Comment (cczhu) [inline baseband/gsb/base.py]:
```
complex_data : bool, optional
    Whether data is complex.  Default: `False`.
```

File: baseband/mark4/base.py [hunk 4]
```diff
@@ -24,34 +24,39 @@ class Mark4FileReader(VLBIFileBase):
     """Simple reader for Mark 4 files.
 
     Adds ``read_frame`` and ``find_frame`` methods to the VLBI file wrapper.
+
+    Parameters
+    ----------
+    ntrack : int
+        Number of Mark 4 bitstreams.
+    decade : int, or None, optional
```
Comment (cczhu) [inline baseband/mark4/base.py]: In line with new docs, neither this nor `ref_time` are optional.

File: baseband/mark4/base.py [hunk 5]
```diff
@@ -24,34 +24,39 @@ class Mark4FileReader(VLBIFileBase):
...
+        Decade in which the observations were taken.  Can instead pass an
+        approximate `ref_time`.
+    ref_time : `~astropy.time.Time`, or None, optional
+        Reference time within 4 years of the observation time.  Used only
+        if `decade` is ``None``.
```
Comment (cczhu) [inline baseband/mark4/base.py]: ``None`` -> `None` (the former doesn't link to the Python core docs).

File: baseband/gsb/base.py [hunk 6]
```diff
@@ -56,31 +56,38 @@ class GSBFileReader(VLBIFileBase):
...
         -------
         frame : `~baseband.gsb.GSBPayload`
             With a ``.data`` property that returns the data encoded.
         """
-        return GSBPayload.fromfile(self.fh_raw, payloadsize=payloadsize,
-                                   nchan=nchan, bps=bps,
-                                   complex_data=complex_data)
+        return GSBPayload.fromfile(self.fh_raw, payloadsize=self.payloadsize,
+                                   nchan=self.nchan, bps=self.bps,
+                                   complex_data=self.complex_data)
```
Comment (cczhu) [inline baseband/gsb/base.py]: Okay, I see the appeal of this now!

File: baseband/mark4/base.py [hunk 7]
```diff
@@ -158,63 +176,59 @@ def determine_ntrack(self, maximum=None):
...
+        first ``ntrack`` for which ``find_frame`` is successful, setting
+        the file's ``ntrack`` property appropriately, and leaving the
         file pointer at the start of the frame.
 
         Parameters
         ----------
         maximum : int, optional
             Maximum number of bytes forward to search through.
-            Default is the framesize (20000 * ntrack // 8).
+            Default is twice the framesize (20000 * ntrack // 8).
```
Comment (cczhu) [inline baseband/mark4/base.py]: `Default: twice the framesize...`

File: baseband/mark4/base.py [hunk 8]
```diff
@@ -158,63 +176,59 @@ def determine_ntrack(self, maximum=None):
...
-            ambiguity in the Mark 4 time stamp.  Required if
-            ``template_header`` is ``None``.
-        maximum : int, optional
-            Maximum number of bytes to search through.  Default is twice the
-            framesize.
         forward : bool, optional
             Seek forward if ``True`` (default), backward if ``False``.
+        maximum : int, optional
+            Maximum number of bytes forward to search through.
+            Default is twice the framesize (20000 * ntrack // 8).
```
Comment (cczhu) [inline baseband/mark4/base.py]: `Default: twice the framesize...`

File: baseband/mark4/base.py [hunk 9]
```diff
@@ -158,63 +176,59 @@ def determine_ntrack(self, maximum=None):
...
 
         Returns
         -------
         header : :class:`~baseband.mark4.Mark4Header`, or None
```
Comment (cczhu) [inline baseband/mark4/base.py]: In line with new docstrings, remove comma before `or None`.

File: baseband/mark5b/base.py [hunk 10]
```diff
@@ -21,74 +21,70 @@ class Mark5BFileReader(VLBIFileBase):
...
+    nchan : int
+        Number of channels encoded in the payload (default=1).
+    bps : int
+        Bits per sample (default=2).
+    kday : int
+        Explicit thousands of MJD of the observation time.  Can instead
+        pass an approximate `ref_time`.
+    ref_time : `~astropy.time.Time`
+        Reference time within 500 days of the observation time, used to
+        infer the full MJD.  Used only if `kday` is ``None``.
```
Comment (cczhu) [inline baseband/mark5b/base.py]: `is not given`

File: baseband/mark5b/base.py [hunk 11]
```diff
@@ -21,74 +21,70 @@ class Mark5BFileReader(VLBIFileBase):
     """Simple reader for Mark 5B files.
 
     Adds ``read_frame`` and ``find_header`` methods to the VLBI file wrapper.
-    """
 
-    def read_frame(self, nchan, bps=2, kday=None, ref_time=None):
+    Parameters
+    ----------
+    nchan : int
```
Comment (cczhu) [inline baseband/mark5b/base.py]:
```
nchan : int, optional
    Number of channels (default: 1).
```

File: baseband/mark5b/base.py [hunk 12]
```diff
@@ -21,74 +21,70 @@ class Mark5BFileReader(VLBIFileBase):
...
+        Number of channels encoded in the payload (default=1).
+    bps : int
```
Comment (cczhu) [inline baseband/mark5b/base.py]:
```
bps : int, optional
    Bits per elementary sample (default = 2).
```

File: baseband/mark5b/base.py [hunk 13]
```diff
@@ -21,74 +21,70 @@ class Mark5BFileReader(VLBIFileBase):
...
-            Explicit thousands of MJD of the observation time, used to infer
-            the full MJD.  If `template_header` and `kday` are both ``None``,
-            any header returned will have its `kday` set to ``None``.
-        maximum : int, optional
-            Maximum number of bytes to search through.  Default is twice the
-            framesize.
         forward : bool, optional
             Seek forward if ``True`` (default), backward if ``False``.
+        maximum : int, optional
+            Maximum number of bytes to search through.  Default is twice the
```
Comment (cczhu) [inline baseband/mark5b/base.py]: `Default: twice the...`

File: baseband/mark5b/base.py [hunk 14]
```diff
@@ -21,74 +21,70 @@ class Mark5BFileReader(VLBIFileBase):
...
+            framesize of 10016 bytes.
 
         Returns
         -------
         header : :class:`~baseband.mark5b.Mark5BHeader`, or None
             Retrieved Mark 5B header, or ``None`` if nothing found.
         """
+        framesize = 10016  # This is fixed for Mark 5B.
```
Comment (cczhu) [inline baseband/mark5b/base.py]: I feel like if this is fixed it should be a class attribute of the frame.

File: baseband/mark5b/tests/test_mark5b.py [hunk 15]
```diff
@@ -203,11 +203,11 @@ def test_payload_getitem_setitem(self, item):
         assert payload2 == payload
 
     def test_frame(self, tmpdir):
-        with mark5b.open(SAMPLE_FILE, 'rb') as fh:
+        with mark5b.open(SAMPLE_FILE, 'rb', nchan=8, bps=2, kday=56000) as fh:
```
Comment (cczhu) [inline baseband/mark5b/tests/test_mark5b.py]: `kday` comes before `nchan` and `bps` now.

File: baseband/mark5b/tests/test_mark5b.py [hunk 16]
```diff
@@ -260,7 +260,7 @@ def test_frame(self, tmpdir):
         assert np.all(frame9.payload.words == 0x11223344)
 
     def test_header_times(self):
-        with mark5b.open(SAMPLE_FILE, 'rb') as fh:
+        with mark5b.open(SAMPLE_FILE, 'rb', nchan=8, bps=2, kday=56000) as fh:
```
Comment (cczhu) [inline baseband/mark5b/tests/test_mark5b.py]: As above, swap `kday` position to just before `nchan`.

**Comment by mhvk** [comment]
Now rebased and ready for review!

**Comment by mhvk** [comment]
Oops, we worked in parallel - But I think I got most of your comments...

File: baseband/mark4/base.py [hunk 17]
```diff
@@ -23,35 +23,41 @@
...
+        Decade in which the observations were taken.  Can instead pass an
+        approximate `ref_time`.
+    ref_time : `~astropy.time.Time` or None
+        Reference time within 4 years of the observation time.  Used only
+        if `decade` is not given.
+    ntrack : int or None, optional.
+        Number of Mark 4 bitstreams.  Can be determined automatically as
+        part of locating the first frame.
     """
+    def __init__(self, fh_raw, decade=None, ref_time=None, ntrack=None):
```
Comment (cczhu) [inline baseband/mark4/base.py]: As with the stream reader, `ntrack` should come before `decade` and `ref_time`.

Though, the original reason why I had that order was for back compatibility with `read_frame(ntrack, decade)`.  Now that we've changed `read_frame`, perhaps it makes more sense to have `sample_rate, decade, ref_time, ntrack, squeeze...` across Mark 4 classes?

File: baseband/mark4/base.py [hunk 18]
```diff
@@ -63,27 +69,40 @@ def find_frame(self, ntrack, maximum=None, forward=True):
...
-        ntrack : int
-            Number of Mark 4 bitstreams.
-        maximum : int, optional
-            Maximum number of bytes forward to search through.
-            Default: twice the framesize (20000 * ntrack // 8).
         forward : bool, optional
             Whether to search forwards or backwards.  Default: `True`.
+        maximum : int, optional
+            Maximum number of bytes forward to search through.
+            Default is twice the framesize (of 20000 * ntrack // 8).
```
Comment (cczhu) [inline baseband/mark4/base.py]: `Default: twice the framesize...`

File: baseband/mark4/base.py [hunk 19]
```diff
@@ -63,27 +69,40 @@ def find_frame(self, ntrack, maximum=None, forward=True):
...
 
         Returns
         -------
-        offset : int or None
-            Byte offset of the next frame.  `None` if the search was not
+        offset : int, or `None`
```
Comment (cczhu) [inline baseband/mark4/base.py]: `offset : int or None`

File: baseband/mark4/base.py [hunk 20]
```diff
@@ -158,63 +177,59 @@ def determine_ntrack(self, maximum=None):
...
+        first ``ntrack`` for which ``find_frame`` is successful, setting
+        the file's ``ntrack`` property appropriately, and leaving the
         file pointer at the start of the frame.
 
         Parameters
         ----------
         maximum : int, optional
             Maximum number of bytes forward to search through.
-            Default: the framesize (20000 * ntrack // 8).
+            Default is twice the framesize (of 20000 * ntrack // 8).
```
Comment (cczhu) [inline baseband/mark4/base.py]: `Default: twice the framesize...`

File: baseband/mark4/base.py [hunk 21]
```diff
@@ -158,63 +177,59 @@ def determine_ntrack(self, maximum=None):
...
-            and `decade` are both not given, any header returned will have its
-            `decade` set to `None`.
-        maximum : int, optional
-            Maximum number of bytes to search through.  Default: twice the
-            framesize.
         forward : bool, optional
             Seek forward if `True` (default), backward if `False`.
+        maximum : int, optional
+            Maximum number of bytes forward to search through.
+            Default is twice the framesize (of 20000 * ntrack // 8).
```
Comment (cczhu) [inline baseband/mark4/base.py]: `Default: twice the framesize...`

File: baseband/mark4/base.py (hunk 17 above)
Comment (cczhu) [reply baseband/mark4/base.py]: Ahh, you did change it in the stream reader.  In that case, the order should change across stream classes too.

File: baseband/mark4/base.py [hunk 22]
```diff
@@ -297,34 +311,30 @@ class Mark4StreamReader(Mark4StreamBase, VLBIStreamReaderBase,
...
+        # Get binary file reader.
+        fh_raw = Mark4FileReader(fh_raw, decade=decade, ref_time=ref_time,
+                                 ntrack=ntrack)
+        # Find first header, determining ntrack if needed.
+        header0 = fh_raw.find_header()
+        assert header0 is not None, (
+            "could not find a first frame using ntrack={}. Perhaps "
+            "try ntrack=None for auto-determination.".format(ntrack))
+        self._offset0 = fh_raw.tell()
+        # Keep possibly auto-determined ntrack, as super will delete it.
```
Comment (cczhu) [inline baseband/mark4/base.py]: Not sure what this comment refers to.

**Comment by cczhu** [comment]
Looks like the docs migrated mostly properly.  Only significant change is possibly shifting `ntrack` around in the Mark 4 frame.py classes to be consistent with base.py (though since it becomes more necessary in the header class it's fine to keep as is as well).

**Comment by mhvk** [comment]
Upon further thought, I've moved `ntrack` back to where it was, mostly because in the end it seemed more important for the binary reader than decade/ref_time -- without it one cannot read files. It also seemed to make sense together with `sample_rate`, something that is absolutely required, but can be auto-determined.

**Comment by cczhu** [comment]
That's fine with me - I don't think there's a perfect solution to `ntrack`.

We're good to merge, I think.

**Comment by mhvk** [comment]
Agreed that there may not be a perfect solution - arguably, it belongs right after `fh_raw`! But I think it is fine now, so will merge.

**Pull request merged**